### PR TITLE
ci: Migrate CI to GitHub Actions workflows

### PR DESCRIPTION
**Pull request by matthewfeickert**
* Closes PR #187
* Requires PR #189

This PR migrates CI from Travis CI and Appveyor to use GitHub Actions workflows. The GHA CI will run across Ubuntu, MacOS, and Windows environments across CPython runtimes spanning 3.6 to 3.9. To reduce the number of runs (especially on slower runners like MacOS) the test matrix only runs on MacOS and Windows for the edge CPython versions: Python 3.6 and Python 3.9. The CI runs on a variety of event triggers:
- Pushes to the `master` branch (PR merges trigger "push" events)
- Pushes to pull requests
- As a nightly CRON job (useful for being alerted to dependencies breaking APIs)
- On demand manual triggers

Travis CI and Appveyor are dropped in this PR and coverage reporting is switched over to use the Codecov GHA (this will require some follow up from the maintainers as you'll want to get an optional `CODECOV_TOKEN` to _greatly_ speed up reporting).

### CODE REVIEW

**Comment by matthewfeickert** [comment]
For demonstration purposes, this PR currently cherry-picks the contents of PR #187 to show that PR #187 is required to get the CI passing. c.f. the following screenshot (the CI runs on my fork until it is merged).

necessary

Once PR #187 is merged this PR will be rebased on top of it to simplify the commit history.

This PR will be WIP until then as there will be some cleaning and simplifying to the README and configs that can easily follow before review.

**Comment by jzuhone** [comment]
@matthewfeickert i merged PR #187.

**Comment by neutrinoceros** [comment]
Hey @matthewfeickert, thanks a lot for taking over this ! Do you need anything to get it across the finish line ?

**Comment by matthewfeickert** [comment]
> Hey @matthewfeickert, thanks a lot for taking over this ! Do you need anything to get it across the finish line ?

:wave: Hi! Yeah, as mentioned in the PR body PR #189 (which is ready for review) should go in first so that the CI actually passes. Once that is in I'll rebase this and get it cleaned up.

**Comment by matthewfeickert** [comment]
@jzuhone @neutrinoceros @matthewturk this is ready for review. Happy to answer any questions on this.

(@neutrinoceros thanks for merging PR #189 :+1: )

File: tox.ini
Comment (matthewfeickert) [inline tox.ini:23]: Not sure exactly what was having problems with `coverage`, but having `coverage<5.0` was causing `matplotlib` and `numpy` to get into fights and have tox try to step backwards through releases until they failed out.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml:52]: For reasonable reporting times you're going to need to create a Codecov token. Once that is made this can get uncommented.

File: tox.ini
Comment (matthewfeickert) [reply tox.ini:23]: Actually, this is still happening on: https://github.com/yt-project/unyt/actions/runs/1105527314

Can someone with admin rights cancel that workflow?

Comment (neutrinoceros) [reply tox.ini:23]: done

File: .github/workflows/ci.yml
Comment (neutrinoceros) [inline .github/workflows/ci.yml:11]: I think this is supposed to enable a button to run the workflow manually, but I'm not sure why we'd need this on top of the other rules (also I don't know where the button is supposed to be found)

File: tox.ini
Comment (matthewfeickert) [reply tox.ini:23]: Can you also get this one: https://github.com/yt-project/unyt/actions/runs/1105527310

File: .github/workflows/ci.yml
Comment (neutrinoceros) [inline .github/workflows/ci.yml:31]: I'm not 100% sure this is valid syntax but seems to be worth a try.
```suggestion
          - os: macos-latest
            python-version: [3.7, 3.8]
          - os: windows-latest
            python-version: [3.7, 3.8]
```

File: tox.ini
Comment (neutrinoceros) [inline tox.ini:94]: This looks like an inconsistency to me. What's the actual minimal supported Python version in unyt atm ? 3.5, 3.6 ?

**Review by neutrinoceros — COMMENTED** [review]
nothing too bad but I want to check this again later before I formally approve. In any case this looks like a solid start, thanks a lot for championing this important task !

File: tox.ini
Comment (neutrinoceros) [reply tox.ini:23]: you got it

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml:6]: Oops this bled in. I need to remove this.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml:11]: I would highly suggest that you keep it. It is useful.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml:31]: Sadly, it is not. :upside_down_face: I wish it was!

File: tox.ini
Comment (matthewfeickert) [reply tox.ini:94]: https://github.com/yt-project/unyt/blob/b9c4c21b2e27fd4af9fafa1949dc01296125543c/setup.py#L51

`setup.py` says Python 3.5, but I was going to suggest that you update that to Python 3.6 as you aren't testing it.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml:31]: To be clear: This would still run, but GHA won't properly parse the `[3.7. 3.8]` syntax and then it just runs everything making the excludes pointless.

File: .github/workflows/ci.yml
Comment (neutrinoceros) [reply .github/workflows/ci.yml:11]: It doesn't seem like it actually costs anything so why not.

File: .github/workflows/ci.yml
Comment (neutrinoceros) [reply .github/workflows/ci.yml:31]: Alright thanks for your reply !

File: tox.ini
Comment (neutrinoceros) [reply tox.ini:94]: Can you test it here ? If it still works it seems worth noting, and if it doesn't then we're already past the point where 3.6 is the de facto min requirement, which is also worth knowing :)

Comment (matthewfeickert) [reply tox.ini:94]: Yeah. I've tested it before on other branches on my fork and `tox` fails on it. 

```pytb
================================================================================================= ERRORS ==================================================================================================
______________________________________________________________________________________ ERROR collecting test session ______________________________________________________________________________________
Direct construction of DocTestTextfilePlus has been deprecated, please use DocTestTextfilePlus.from_parent.
See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.

---------- coverage: platform linux, python 3.5.10-final-0 -----------
Name                               Stmts   Miss  Cover
------------------------------------------------------
unyt/__init__.py                      33     33     0%
unyt/_parsing.py                      38     38     0%
unyt/_physical_ratios.py             111    111     0%
unyt/_pint_conversions.py             16     16     0%
unyt/_unit_lookup_table.py            51     51     0%
unyt/array.py                        708    708     0%
unyt/dimensions.py                    88     88     0%
unyt/equivalencies.py                166    166     0%
unyt/exceptions.py                    66     66     0%
unyt/mpl_interface.py                 76     76     0%
unyt/physical_constants.py             3      3     0%
unyt/testing.py                       14     14     0%
unyt/tests/__init__.py                 0      0   100%
unyt/tests/test_define_unit.py        34     34     0%
unyt/tests/test_mpl_interface.py     175    175     0%
unyt/tests/test_unit_registry.py      84     84     0%
unyt/tests/test_unit_systems.py       75     75     0%
unyt/tests/test_units.py             524    524     0%
unyt/tests/test_unyt_array.py       1608   1608     0%
unyt/tests/test_unyt_testing.py       27     27     0%
unyt/unit_object.py                  464    464     0%
unyt/unit_registry.py                157    157     0%
unyt/unit_symbols.py                   8      8     0%
unyt/unit_systems.py                 159    159     0%
------------------------------------------------------
TOTAL                               4685   4685     0%

========================================================================================= short test summary info =========================================================================================
ERROR 
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================ 1 error in 0.56s =============================================================================================
ERROR: InvocationError for command /unyt/.tox/py/bin/pytest --cov=unyt --cov-append --doctest-modules --doctest-plus --doctest-rst --basetemp=/unyt/.tox/py/tmp -W once (exited with code 2)
py finish: run-test  after 1.18 seconds
py start: run-test-post 
py finish: run-test-post  after 0.00 seconds
_________________________________________________________________________________________________ summary _________________________________________________________________________________________________
ERROR:   py: commands failed
cleanup /unyt/.tox/.tmp/package/1/unyt-2.8.0+29.g4cb5ec5.dirty.zip
```

Though I just went and in a local `python:3.5` Docker image tested things _without_ `tox` and just running `pytest` manually I can get things to pass.

```console
(venv) root@4bb7b37a5d66:/unyt# python --version --version
Python 3.5.10
(venv) root@4bb7b37a5d66:/unyt# pytest
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.5.10, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /unyt, configfile: setup.cfg
plugins: cov-2.12.1, doctestplus-0.5.0
collected 136 items                                                                                                                                                                                       

unyt/tests/test_define_unit.py ..                                                                                                                                                                   [  1%]
unyt/tests/test_linters.py ..                                                                                                                                                                       [  2%]
unyt/tests/test_mpl_interface.py ...............                                                                                                                                                    [ 13%]
unyt/tests/test_unit_registry.py .......                                                                                                                                                            [ 19%]
unyt/tests/test_unit_systems.py .......                                                                                                                                                             [ 24%]
unyt/tests/test_units.py ................................................                                                                                                                           [ 59%]
unyt/tests/test_unyt_array.py ..................................................                                                                                                                    [ 96%]
unyt/tests/test_unyt_testing.py .....                                                                                                                                                               [100%]

========================================================================================== 136 passed in 11.76s ===========================================================================================
(venv) root@4bb7b37a5d66:/unyt#
```

So I assume that there's some tox problems that I can debug.

**Comment by ngoldbaum** [comment]
It’s probably not worth the effort to keep python 3.5 working, it’s EOL.

**Comment by matthewfeickert** [comment]
> It’s probably not worth the effort to keep python 3.5 working, it’s EOL.

This would be my preferred option.

**Review by jzuhone — APPROVED** [review]

**Comment by matthewfeickert** [comment]
I rebased this and force pushed just to clean things up a bit, but nothing has changed since it was reviewed.

**Comment by neutrinoceros** [comment]
Siding with Nathan here: not worth maintaining compatibility with an EOL version of Python. Can you simply drop every mention of Python 3.5 (in particular in the tox config) ? I'll approve once this is done.

**Comment by matthewfeickert** [comment]
> Siding with Nathan here: not worth maintaining compatibility with an EOL version of Python. Can you simply drop every mention of Python 3.5 (in particular in the tox config) ?

Sure. `tox` is clean now:

```console
$ git grep "35" tox.ini
```

returns nothing.

> It would also be nice to update the relevant metadata and requirements in `setup.py`.

You want this all in this PR? Given how much information on Python 3.5 is spread across the repository I'd suggest just making a follow up PR as docs stuff is out of scope for a CI PR and is making things less atomic IMO.

```console
$ git grep "3\.5" ':!unyt/tests/*'
CONTRIBUTING.rst:    $ pyenv install -s 3.5.9
CONTRIBUTING.rst:    $ pyenv local 3.5.9 3.6.10 3.7.6 3.8.1
CONTRIBUTING.rst:   $ pyenv local 3.5.9 3.6.10 3.7.6 3.8.1
CONTRIBUTING.rst:3. The pull request should work for Python 3.5, 3.6, 3.7, and 3.8. Check in the
HISTORY.rst:* Dropped support for Python 3.5.
docs/usage.rst:  Europa: 3.55 day
docs/usage.rst:  unyt_array([ 1.76926617,  3.55125104,  7.15465486, 16.68926145], 'day')
setup.py:        "Programming Language :: Python :: 3.5",
setup.py:    python_requires=">=3.5",
unyt/equivalencies.py:    >>> (3.5*g).to_equivalent('J', 'mass_energy')
versioneer.py:* Compatible With: python2.6, 2.7, 3.2, 3.3, 3.4, 3.5, 3.6, and pypy
```

If you do I'll do it, but thought I'd ask as I don't mind making a follow up PR.

**Comment by neutrinoceros** [comment]
I only asked within the assumption that it was a small task. If it's not, I agree that a follow up PR will do :)

**Review by neutrinoceros — APPROVED** [review]

**Comment by jzuhone** [comment]
I think we should give @ngoldbaum (or anyone else) a chance to do another look, but since it's the weekend I will wait to merge Tuesday morning Eastern Time in the US if there are no comments by then.

**Comment by matthewfeickert** [comment]
> I agree that a follow up PR will do :)

This is now PR #194 (which can be rebased after this is merged). :+1:

**Pull request merged**
